File: intv/TopKMovies.cpp
#include <bits/stdc++.h>
using namespace std;

// Movie class
class Movie {
private:
    string title;
    int views;

public:
    Movie(string t) : title(t), views(0) {}

    string getTitle() const { return title; }
    int getViews() const { return views; }

    // Increment views when played
    void play() { views++; }

    // Comparator for set
    bool operator<(const Movie &other) const {
        if (views == other.views)
            return title < other.title;
        return views > other.views;
    }
};

// MovieService class
class MovieService {
private:
    unordered_map<string, Movie> movieMap; // fast lookup by title
    set<Movie> moviesByViews;               // sorted by views

public:
    // Add a new movie
    void addMovie(const string &title) {
        Movie m(title);
        movieMap.insert({title, m});
        moviesByViews.insert(m);
    }

    // Play a movie by title
    void playMovie(const string &title) {
        auto it = movieMap.find(title);
        if (it == movieMap.end()) {
            cout << "Movie not found: " << title << endl;
            return;
        }

        // Update views and maintain set ordering
        Movie updated = it->second;
        moviesByViews.erase(updated);
        updated.play();
        moviesByViews.insert(updated);
        it->second = updated;
    }

    // Print all movies sorted by views
    void printMovies() const {
        cout << "Movies sorted by views:\n";
        for (const auto &m : moviesByViews) {
            cout << m.getTitle() << " (" << m.getViews() << " views)\n";
        }
    }
};

int main() {
    MovieService service;

    // Add movies
    service.addMovie("Inception");
    service.addMovie("Matrix");
    service.addMovie("Tenet");
    service.addMovie("Avatar");

    // Simulate playing movies
    service.playMovie("Inception");
    service.playMovie("Inception");
    service.playMovie("Matrix");
    service.playMovie("Avatar");
    service.playMovie("Avatar");
    service.playMovie("Avatar");

    // Print results
    service.printMovies();

    return 0;
}

File: intv/TopKMovies.java
import java.util.*;

class Movie implements Comparable<Movie> {
    private String title;
    private int views;

    public Movie(String title) {
        this.title = title;
        this.views = 0;
    }

    public String getTitle() {
        return title;
    }

    public int getViews() {
        return views;
    }

    public void play() {
        views++;
    }

    // Comparator for TreeSet: descending views, tie-breaker: title ascending
    @Override
    public int compareTo(Movie other) {
        if (this.views != other.views) {
            return Integer.compare(other.views, this.views); // descending
        }
        return this.title.compareTo(other.title); // ascending
    }

    // Required for TreeSet remove() to work properly
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Movie)) return false;
        Movie movie = (Movie) o;
        return title.equals(movie.title);
    }
}

class MovieService {
    private Map<String, Movie> movieMap = new HashMap<>();
    private TreeSet<Movie> moviesByViews = new TreeSet<>();

    public void addMovie(String title) {
        Movie m = new Movie(title);
        movieMap.put(title, m);
        moviesByViews.add(m);
    }

    public void playMovie(String title) {
        Movie m = movieMap.get(title);
        if (m == null) {
            System.out.println("Movie not found: " + title);
            return;
        }

        // Update TreeSet ordering
        moviesByViews.remove(m);
        m.play();
        moviesByViews.add(m);
    }

    public void printMovies() {
        System.out.println("Movies sorted by views:");
        for (Movie m : moviesByViews) {
            System.out.println(m.getTitle() + " (" + m.getViews() + " views)");
        }
    }
}

class Main {
    public static void main(String[] args) {
        MovieService service = new MovieService();

        // Add movies
        service.addMovie("Inception");
        service.addMovie("Matrix");
        service.addMovie("Tenet");
        service.addMovie("Avatar");

        // Simulate playing movies
        service.playMovie("Inception");
        service.playMovie("Inception");
        service.playMovie("Matrix");
        service.playMovie("Avatar");
        service.playMovie("Avatar");

        // Print results
        service.printMovies();
    }
}
